src/iarc7_sensors/orientation_filter.py

```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python2
+
+from iarc7_safety.SafetyClient import SafetyClient
+import math
+import rospy
+import tf.transformations
+import tf2_ros
+import threading
+
+from geometry_msgs.msg import TransformStamped
+from iarc7_msgs.msg import (Float64Stamped,
+                            OrientationAnglesStamped)
+
+class OrientationFilter(object):
+    def __init__(self):
+        rospy.init_node('orientation_filter')
+
+        self._lock = threading.Lock()
+
+        self._orientation_sub = rospy.Subscriber(
+                'fc_orientation',
+                OrientationAnglesStamped,
+                lambda msg: self._callback(OrientationAnglesStamped, msg))
+
+        # This topic is assumed to have rotation around +z
+        # (angle from +x with +y being +pi/2)
+        self._line_yaw_sub = rospy.Subscriber(
+                'line_yaw',
+                Float64Stamped,
+                lambda msg: self._callback(Float64Stamped, msg))
+
+        self._line_weight = rospy.get_param('~line_weight')
+        self._message_queue_length = rospy.get_param('~message_queue_length')
+
+        initial_msg = Float64Stamped()
+        initial_msg.header.stamp = rospy.Time()
+        initial_msg.data = rospy.get_param('~initial_yaw', float('NaN'))
+
+        # This queue contains pairs, sorted by timestamp, oldest at index 0
+        #
+        # First item of each pair is a 3-tuple containing:
+        #     yaw (rotation around +z)
+        #     pitch (rotation around +y')
+        #     roll (rotation around +x'')
+        # Second item of each pair is the message itself
+        #
+        # The yaw in the 3-tuple is the current value of the filter at that
+        # point, the pitch and roll are too except that they're simply
+        # propogated forward from the last OrientationAnglesStamped message
+        self._queue = [([initial_msg.data, float('NaN'), float('NaN')],
+                        initial_msg)]
+
+        self._transform_broadcaster = tf2_ros.TransformBroadcaster()
+
+        self._last_published_stamp = rospy.Time(0)
+
+    def _publish_transform(self, r, p, y, time):
+        transform_msg = TransformStamped()
+
+        transform_msg.header.stamp = time
+        transform_msg.header.frame_id = 'level_quad'
+        transform_msg.child_frame_id = 'quad'
+
+        quaternion = tf.transformations.quaternion_from_euler(r, p, y)
+        transform_msg.transform.rotation.x = quaternion[0]
+        transform_msg.transform.rotation.y = quaternion[1]
+        transform_msg.transform.rotation.z = quaternion[2]
+        transform_msg.transform.rotation.w = quaternion[3]
+
+        self._transform_broadcaster.sendTransform(transform_msg)
+
+    def _get_last_index(self, klass, time=None):
+        '''
+        Return the index of the last message of type `klass` (or the
```

Comment:
This comment is misleading. It returns the last message of type 'klass' that is also before 'time'. If time is unspecified it is the last message of type 'klass'